### PR TITLE
Added last-image option for big-bang.

### PR DESCRIPTION
**Pull request by ds26gte**
Current solution to brownplt/pyret-lang#808. Previous solution(s) and discussion in #180.

### CODE REVIEW

File: src/web/js/trove/world-lib.js [hunk 1]
```diff
@@ -750,10 +759,20 @@
                 k2();
               } else {
                 if (extras.closeWhenStop) {
-                  if (extras.closeBigBangWindow) {
-                    extras.closeBigBangWindow();
+                  if (lif) {
+                    runtime.ffi.throwMessageException('Cannot have both last-image and close-when-stop(true)');
```
Comment (jpolitz) [inline src/web/js/trove/world-lib.js]: Two things – if there's a dynamic error running a world, it should probably still shut down the world.  Also, why even _start_ the process running if there's an error?  Can't we just warn up front that this combination is disallowed?

File: src/web/js/trove/world-lib.js [hunk 2]
```diff
@@ -750,10 +759,20 @@
...
+                  } else {
+                    if (extras.closeBigBangWindow) {
+                      extras.closeBigBangWindow();
+                    }
+                    Jsworld.shutdownSingle({cleanShutdown: true});
                   }
-                  Jsworld.shutdownSingle({cleanShutdown: true});
+                } else if (lif) {
+                  var handler = lif();
```
Comment (jpolitz) [inline src/web/js/trove/world-lib.js]: Which of these lines call the user's Pyret code?  There's probably some need for stack management here.

File: src/web/js/trove/world-lib.js (hunk 1 above)
Comment (ds26gte) [inline src/web/js/trove/world-lib.js]: Here's the program I used (a blinkered rewrite of a Finnish WeScheme program):

```
import image as I
import world as W

var TURTLE = I.isosceles-triangle(20, 50, 'solid', 'black')

var ORIGIN = I.circle(3, 'solid', 'red')

var WIDTH = 500

var HEIGHT = 500

var COLOR = 'blue'

var BG-COLOR = 'white'

var PEN-CAP = 'round'

var PEN-JOIN = 'round'

var PEN-STYLES = [list: 'solid', 'dot', 'long-dash', 'short-dash', 'dot-dash']

fun make-position(x, y, angle):
  { x: x,
    y: y,
    angle: angle
  }
end

fun make-turtle(location, mirror-x, mirror-y):
  { location: location,
    mirror-x: mirror-x,
    mirror-y: mirror-y
  }
end

fun make-animation(location, image, commands, color, penP, pen-size, pen-style, bg-color,
                   stampP, stamps, origin, turtleP):
  {
    location: location,
    image: image,
    commands: commands,
    color: color,
    penP: penP,
    pen-size: pen-size,
    pen-style: pen-style,
    bg-color: bg-color,
    stampP: stampP,
    stamps: stamps,
    origin: origin,
    turtleP: turtleP
  }
end

fun listP(x):
  is-empty(x) or is-link(x)
end

fun get-position(anim):
  anim.location.location
end

fun get-x(anim):
  anim.location.location.x
end

fun get-y(anim):
  anim.location.location.y
end

fun get-origin-x(anim):
  anim.origin.x
end

fun get-origin-y(anim):
  anim.origin.y
end

fun get-height(anim):
  I.image-height(anim.image)
end

fun get-width(anim):
  I.image-width(anim.image)
end

fun get-angle(anim):
  anim.location.location.angle
end

fun get-mirror-x(anim):
  anim.location.mirror-x
end

fun get-mirror-y(anim):
  anim.location.mirror-y
end

fun get-color(anim):
  if I.is-image-color(anim.color) or is-string(anim.color):
    anim.color
  else if is-link(anim.color):
    anim.color.first
  else:
    'black'
  end
end

fun t-get-x(t):
  t.location.x
end

fun t-get-y(t):
  t.location.y
end

fun t-get-angle(t):
  t.location.angle
end

fun on-screenP(pos, img):
  (0 <= pos.x) and (pos.x <= I.image-width(img)) and
  (0 <= pos.y) and (pos.y <= I.image-height(img))
end

pi = 4 * num-atan(1)

fun degrees-to-radians(deg):
  (deg * pi) / 180
end

fun draw-horizontal-line(y, img, color):
  I.add-line(img, 0, y, I.image-width(img), y, color)
end

fun draw-vertical-line(x, img, color):
  I.add-line(img, x, 0, x, I.image-height(img), color)
end

fun draw-vertical-lines(x-step, orig, img, color):
  let start-x = num-modulo(orig.x, x-step),
      times = num-round(I.image-width(img) / x-step),
      steps = range(0, times).
              map(lam(x): start-x + (x * x-step) end):
    steps.foldl(lam(x, i): draw-vertical-line(x, i, color) end, img)
  end
end

fun draw-horizontal-lines(y-step, orig, img, color):
  let start-y = num-modulo(orig.y, y-step),
      times = num-round(I.image-height(img) / y-step),
      steps = range(0, times).
              map(lam(y): start-y + (y * y-step) end):
    steps.foldl(lam(y, i): draw-horizontal-line(y, i, color) end, img)
  end
end

fun draw-grid(x-step, y-step, origin, img, color):
  draw-vertical-lines(x-step, origin,
    draw-horizontal-lines(y-step, origin, img, color), color)
end

fun new-turtle-position(anim, dist):
  let x = get-x(anim),
      y = get-y(anim),
      a = get-angle(anim):
    if dist == 0:
      make-position(x, y, a)
    else if (a == 0) or (a == 360):
      make-position(x, y - dist, a)
    else if a == 90:
      make-position(x - dist, y, a)
    else if a == 180:
      make-position(x, y + dist, a)
    else if a == 270:
      make-position(x + dist, y, a)
    else if (0 < a) and (a < 90):
      make-position(x - (dist * num-cos(degrees-to-radians(90 - a))),
                    y - (dist * num-sin(degrees-to-radians(90 - a))), a)
    else if (90 < a) and (a < 180):
      make-position(x - (dist * num-cos(degrees-to-radians(a - 90))),
                    y + (dist * num-sin(degrees-to-radians(a - 90))), a)
    else if (180 < a) and (a < 270):
      make-position(x + (dist * num-cos(degrees-to-radians(270 - a))),
                    y + (dist * num-sin(degrees-to-radians(270 - a))), a)
    else if (270 < a) and (a < 360):
      make-position(x + (dist * num-cos(degrees-to-radians(a - 270))),
                    y - (dist * num-sin(degrees-to-radians(a - 270))), a)
    else:
      make-position(x, y, a)
    end
  end
end

fun new-turtle-direction(anim, turn):
  let x = get-x(anim),
      y = get-y(anim),
      a = get-angle(anim),
      turning-angle = num-modulo(num-round(turn), 360):
    if 0 == turning-angle:
      make-position(x, y, a)
    else if 0 < turning-angle:
      make-position(x, y, num-modulo(a + turning-angle, 360))
    else if (0 > turning-angle) and (num-abs(turning-angle) > a):
      make-position(x, y, 360 + num-modulo(a + turning-angle, 360))
    else if (0 > turning-angle) and (num-abs(turning-angle) <= a):
      make-position(x, y, a + turning-angle)
    else:
      make-position(x, y, a)
    end
  end
end

fun new-turtle-direction-full(anim, turn):
  set-location(anim, make-turtle(new-turtle-direction(anim, turn),
                                 get-mirror-x(anim), get-mirror-y(anim)))
end

fun new-turtle-position-full(anim, turn):
  set-location(anim, make-turtle(new-turtle-position(anim, turn),
                                 get-mirror-x(anim), get-mirror-y(anim)))
end

fun set-location(anim, turt):
  anim.{ location: turt }
end

fun set-angle(anim, a):
  if is-number(a):
    set-location(anim, make-turtle(make-position(get-x(anim), get-y(anim), a),
                                   get-mirror-x(anim), get-mirror-y(anim)))
  else:
    anim
  end
end

fun set-image(anim, img):
  if I.is-image(img):
    anim.{ image: I.overlay(img, anim.image) }
  else:
    anim
  end
end

fun set-pen(anim, setting):
  if is-boolean(setting):
    anim.{ penP: setting }
  else:
    anim
  end
end

fun set-color(anim, setting):
  if is-string(setting) or I.is-image-color(setting) or
    (listP(setting) and setting.all(lam(x): is-string(x) or I.is-image-color(x) end)):
    anim.{ color: setting }
  else:
    anim
  end
end

fun set-bg-color(anim, setting):
  if is-string(setting) or I.is-image-color(setting):
    anim.{
      image: I.overlay(I.rectangle(get-width(anim), get-height(anim), 'solid', setting),
                       anim.image),
      bg-color: setting
    }
  else:
    anim
  end
end

fun x-insideP(x, origon-x, width):
  ((0 - origon-x) <= x) and (x <= (width - origon-x))
end

fun y-insideP(y, origon-y, height):
  ((0 - (height - origon-y)) <= y) and (y <= origon-y)
end

fun set-position(anim, x, y):
  if x-insideP(x, get-origin-x(anim), get-width(anim)) and
     y-insideP(y, get-origin-y(anim), get-height(anim)):
    set-location(anim, make-turtle(make-position(get-origin-x(anim) + x,
                                                 get-origin-y(anim) - y,
                                                 get-angle(anim)),
                                   get-mirror-x(anim),
                                   get-mirror-y(anim)))
  else:
    anim
  end
end

fun clean-up-image(anim):
  set-image(anim, I.rectangle(get-width(anim), get-height(anim), 'solid',
                              anim.bg-color))
end

fun set-stamper-image(anim, img):
  if I.is-image(img) or (listP(img) and img.all(I.is-image)):
    anim.{
      stampP: true,
      stamps: img
    }
  else:
    anim
  end
end

fun set-stamper-off(anim):
  anim.{
    stampP: false,
    stamps: I.empty-image
  }
end

fun set-mirror-y-on(anim):
  set-location(anim, make-turtle(get-position(anim), get-x(anim), get-mirror-y(anim)))
end

fun set-mirror-y-off(anim):
  set-location(anim, make-turtle(get-position(anim), false, get-mirror-y(anim)))
end

fun set-mirror-x-on(anim):
  set-location(anim, make-turtle(get-position(anim), get-mirror-x(anim), get-y(anim)))
end

fun set-origin-xy(anim):
  anim.{
    origin: make-position(get-x(anim), get-y(anim), 0)
  }
end

fun set-pen-size(anim, size):
  if (0 <= size) and (size <= 255):
    anim.{
      pen-size: size
    }
  else:
    anim
  end
end

fun is-pen-styleP(str):
  PEN-STYLES.member(str)
end

fun set-pen-style(anim, style):
  if is-pen-styleP(style):
    anim.{
      pen-style: style
    }
  else:
    anim
  end
end

fun set-turtle-on-off(anim, setting):
  if is-boolean(setting):
    anim.{
      turtleP: setting
    }
  else:
    anim
  end
end

fun set-mirror-x-off(anim):
  set-location(anim, make-turtle(get-position(anim), get-mirror-x(anim), false))
end

fun set-and-draw-bg-grid(anim, x-step, y-step, color):
  set-image(anim, draw-grid(x-step, y-step, anim.origin, anim.image, color))
end

fun set-bg-image-for(anim, img):
  set-image(anim, img)
end

fun set-position-to-origin(anim):
  set-location(anim, make-turtle(make-position(get-origin-x(anim), get-origin-y(anim),
                                               get-angle(anim)),
                                 get-mirror-x(anim), get-mirror-y(anim)))
end

fun draw-image(new-pos, old-pos, img, color, size, style):
  if on-screenP(new-pos, img) and on-screenP(old-pos, img):
    I.add-line(img, old-pos.x, old-pos.y, new-pos.x, new-pos.y, color)
  else:
    img
  end
end

fun mirror(p, mp):
  let delta = num-abs(p - mp):
    if p < mp:
      mp + delta
    else:
      mp - delta
    end
  end
end

fun draw-m-image(next-step, anim):
  if next-step.penP:
    let img = draw-image(get-position(next-step), get-position(anim), next-step.image,
                         get-color(next-step), next-step.pen-size, next-step.pen-style),
        mx = get-mirror-x(anim),
        my = get-mirror-y(anim),
        old-x = get-x(anim),
        old-y = get-y(anim),
        new-x = get-x(next-step),
        new-y = get-y(next-step),
        old-angle = get-angle(anim),
        new-angle = get-angle(next-step):
      if is-number(mx) and is-number(my):
        draw-image(make-position(mirror(new-x, mx), mirror(new-y, my), new-angle),
                   make-position(mirror(old-x, mx), mirror(old-y, my), old-angle),
                   img, get-color(next-step), next-step.pen-size, next-step.pen-style)
      else if is-number(mx):
        draw-image(make-position(mirror(new-x, mx), new-y, new-angle),
                   make-position(mirror(old-x, mx), old-y, old-angle),
                   img, get-color(next-step), next-step.pen-size, next-step.pen-style)
      else if is-number(my):
        draw-image(make-position(new-x, mirror(new-y, my), new-angle),
                   make-position(old-x, mirror(old-y, my), old-angle),
                   img, get-color(next-step), next-step.pen-size, next-step.pen-style)
      else:
        img
      end
    end
  else:
    next-step.image
  end
end

fun draw-stamp(posit, img, st):
  if on-screenP(posit, img):
    I.place-image(st, posit.x, posit.y, img)
  else:
    img
  end
end

fun draw-m-stamp(t, img, st):
  let image = draw-stamp(t.location, img, st),
      mx = t.mirror-x,
      my = t.mirror-y,
      new-x = t-get-x(t),
      new-y = t-get-y(t),
      new-angle = t-get-angle(t):
    if is-number(mx) and is-number(my):
      draw-stamp(make-position(mirror(new-x, mx), mirror(new-y, my), new-angle),
                 image, st)
    else if is-number(mx):
      draw-stamp(make-position(mirror(new-x, mx), new-y, new-angle),
                 image, st)
    else if is-number(my):
      draw-stamp(make-position(new-x, mirror(new-y, my), new-angle),
                 image, st)
    else:
      image
    end
  end
end

fun draw-animation(anim):
  if anim.turtleP:
    I.place-image(ORIGIN, get-origin-x(anim), get-origin-y(anim),
                I.place-image(I.rotate(get-angle(anim), TURTLE),
                            get-x(anim), get-y(anim), anim.image))
  else:
    anim.image
  end
end

fun draw-last(anim):
  anim.image
end

fun readyP(anim):
  is-empty(anim.commands)
end

fun handle-automatically(anim):
  handle-command(anim, ' ')
end

fun stationaryP(old-pos, new-pos):
  (old-pos.x == new-pos.x) and (old-pos.y == new-pos.y)
end

fun number-of-stamps(anim):
  if anim.stampP and I.is-image(anim.stamps):
    1
  else if anim.stampP and listP(anim.stamps):
    anim.stamps.length()
  else:
    0
  end
end

fun number-of-colors(anim):
  if anim.penP and (I.is-image-color(anim.color) or is-string(anim.color)):
    1
  else if anim.penP and (listP(anim.color)):
    anim.color.length()
  else:
    0
  end
end

fun handle-command(anim, k):
  if is-empty(anim.commands):
    anim
  else:
    let next-step = anim.commands.first(anim),
        movedP = not(stationaryP(get-position(anim), get-position(next-step))):
      if movedP and (number-of-stamps(next-step) == 1) and
                    (number-of-colors(next-step) == 1):
        next-step.{
          image: draw-m-stamp(next-step.location,
                              draw-m-image(next-step, anim),
                              next-step.stamps),
          commands: anim.commands.rest
        }
      else if movedP and (number-of-stamps(next-step) > 1) and
                         (number-of-stamps(next-step) > 1):
        next-step.{
          image: draw-m-stamp(next-step.location,
                              draw-m-image(next-step, anim),
                              next-step.stamps.first),
          commands: anim.commands.rest,
          color: next-step.color.rest.append([list: next-step.color.first]),
          stamps: next-step.stamps.rest.append([list: next-step.stamps.first])
        }
      else if movedP and (number-of-stamps(next-step) > 1) and
                         (number-of-colors(next-step) == 1):
        next-step.{
          image: draw-m-stamp(next-step.location,
                              draw-m-image(next-step, anim),
                              next-step.stamps.first),
          commands: anim.commands.rest,
          stamps: next-step.stamps.rest.append([list: next-step.stamps.first])
        }
      else if movedP and (number-of-stamps(next-step) == 1) and
                         (number-of-colors(next-step) > 1):
        next-step.{
          image: draw-m-stamp(next-step.location,
                              draw-m-image(next-step, anim),
                              next-step.stamps.first),
          commands: anim.commands.rest,
          color: next-step.color.rest.append([list: next-step.color.first])
        }
      else if movedP and (number-of-stamps(next-step) == 1) and
                         (number-of-colors(next-step) == 0):
        next-step.{
          image: draw-m-stamp(next-step.location,
                              draw-m-image(next-step, anim),
                              next-step.stamps),
          commands: anim.commands.rest
        }
      else if movedP and (number-of-stamps(next-step) == 0) and
                         (number-of-colors(next-step) == 1):
        next-step.{
          image: draw-m-image(next-step, anim),
          commands: anim.commands.rest
        }
      else if movedP and (number-of-stamps(next-step) > 1) and
                         (number-of-colors(next-step) == 0):
        next-step.{
          image: draw-m-stamp(next-step.location, next-step.image, next-step.stamps.first),
          commands: anim.commands.rest,
          stamps: next-step.stamps.rest.append([list: next-step.stamps.first])
        }
      else if movedP and (number-of-stamps(next-step) == 0) and
                         (number-of-colors(next-step) > 1):
        next-step.{
          image: draw-m-image(next-step, anim),
          commands: anim.commands.rest,
          color: next-step.color.rest.append([list: next-step.color.first])
        }
      else:
        next-step.{
          image: draw-m-image(next-step, anim),
          commands: anim.commands.rest
        }
      end #if
    end #let
  end #if
end #fun

fun reverse-flatten-into(x, lst):
  if not(is-function(x)):
    lst.foldl(reverse-flatten-into, x)
  else:
    link(x, lst)
  end
end

fun flatten(lst):
  reverse-flatten-into(lst, empty).reverse()
end

fun forward(distance):
  lam(x):
    new-turtle-position-full(x, distance)
  end
end

fun turn-left(turn):
  lam(x):
    new-turtle-direction-full(x, turn)
  end
end

fun turn-right(turn):
  lam(x):
    new-turtle-direction-full(x, -1 * turn)
  end
end

fun pen-up():
  lam(x):
    set-pen(x, false)
  end
end

fun pen-down():
  lam(x):
    set-pen(x, true)
  end
end

fun change-color(color):
  lam(x):
    set-color(x, color)
  end
end

fun t-repeat(times, commands):
  if times <= 0:
    empty
  else if listP(commands):
    commands.append(t-repeat(times - 1, commands))
  else:
    [list: commands].append(t-repeat(times - 1, commands))
  end
end

fun change-bg-color(color):
  lam(x):
    set-bg-color(x, color)
  end
end

fun go-to(x, y):
  lam(z):
    set-position(z, x, y)
  end
end

fun clean-up():
  clean-up-image
end

fun stamper-on(img):
  lam(x):
    set-stamper-image(x, img)
  end
end

fun stamper-off():
  set-stamper-off
end

fun mirror-x-on():
  set-mirror-x-on
end

fun mirror-y-on():
  set-mirror-y-on
end

fun mirror-x-off():
  set-mirror-x-off
end

fun mirror-y-off():
  set-mirror-y-off
end

fun set-origin():
  set-origin-xy
end

fun set-bg-grid(x, y, color):
  lam(z):
    set-and-draw-bg-grid(z, x, y, color)
  end
end

fun set-bg-image(img):
  lam(x):
    set-bg-image-for(x, img)
  end
end

fun go-to-origin():
  set-position-to-origin
end

fun change-pen-size(size):
  lam(x):
    set-pen-size(x, size)
  end
end

fun change-pen-style(size):
  lam(x):
    set-pen-style(x, size)
  end
end

fun hide-turtle():
  lam(x):
    set-turtle-on-off(x, false)
  end
end

fun show-turtle():
  lam(x):
    set-turtle-on-off(x, true)
  end
end

fun reset-angle():
  lam(x):
    set-angle(x, 0)
  end
end

fun create-animation(list-of-commands, width, height):
  make-animation(make-turtle(make-position(width / 2, height / 2, 0), false, false),
                 I.empty-scene(width, height),
                 flatten(list-of-commands),
                 COLOR, true, 1, 'solid', BG-COLOR, false, empty,
                 make-position(0, height, 0), true)
end

fun give-speed(sp):
  if sp > 0:
    sp
  else:
    1/28
  end
end

fun draw(list-of-commands):
  W.big-bang(create-animation(list-of-commands, WIDTH, HEIGHT),
             [list:
               W.to-draw(draw-animation),
               W.on-tick(handle-automatically),
               W.stop-when(readyP),
               W.last-image(draw-last)]).image
end

fun draw-custom(list-of-commands, width, height, speed):
  W.big-bang(create-animation(list-of-commands, width, height),
             [list:
               W.to-draw(draw-animation),
               W.on-tick-n(handle-automatically, give-speed(speed)),
               W.stop-when(readyP),
               W.last-image(draw-last)]).image
end

fun draw-and-store(list-of-commands):
  W.big-bang(create-animation(list-of-commands, WIDTH, HEIGHT),
             [list:
               W.to-draw(draw-animation),
               W.on-tick(handle-automatically),
               W.stop-when(readyP),
               W.last-image(draw-last)]).image
end

fun draw-and-store-custom(list-of-commands, width, height, speed):
  W.big-bang(create-animation(list-of-commands, width, height),
             [list:
               W.to-draw(draw-animation),
               W.on-tick-n(handle-automatically, give-speed(speed)),
               W.stop-when(readyP),
               W.last-image(draw-last)]).image
end

fun draw-step-by-step(list-of-commands):
  W.big-bang(create-animation(list-of-commands, WIDTH, HEIGHT),
             [list:
               W.on-key(handle-command),
               W.to-draw(draw-animation),
               W.stop-when(readyP),
               W.last-image(draw-last)]).image
end

fun draw-step-by-step-custom(list-of-commands, width, height):
  W.big-bang(create-animation(list-of-commands, width, height),
             [list:
               W.on-key(handle-command),
               W.to-draw(draw-animation),
               W.stop-when(readyP),
               W.last-image(draw-last)]).image
end

fun piirrae(list-of-commands):
  draw(list-of-commands, WIDTH, HEIGHT)
end

fun piirrae-osissa(list-of-commands):
  draw-step-by-step(list-of-commands, WIDTH, HEIGHT)
end

var KUVIO = [list: set-origin(),
                   set-bg-grid(30, 30, 'green'),
                   change-color('black'),
                   go-to(90, 60)]

draw(KUVIO)
```

Comment (jpolitz) [inline src/web/js/trove/world-lib.js]: Do you have any smaller examples?  Should the example I posted work?

File: src/web/js/trove/world-lib.js (hunk 2 above)
Comment (ds26gte) [inline src/web/js/trove/world-lib.js]: The example should work as is. Paste in entirety into the def window, and click Run.

File: src/web/js/trove/world-lib.js (hunk 1 above)
Comment (ds26gte) [inline src/web/js/trove/world-lib.js]: A `.image` at the end?

Comment (ds26gte) [inline src/web/js/trove/world-lib.js]: Sorry. Am involved in updating Patch to the latest horizon. I'll get back to testing your program when I can leave the update at a stable point.

File: src/web/js/trove/world-lib.js (hunk 2 above)
Comment (ds26gte) [inline src/web/js/trove/world-lib.js]: daa7a18 (and the commit before it) includes the fixes. The last-image handler wasn't handing its canvas node on to its wrapper's success continuation. Also, made the handler's user code execute inside a safeCall.

Here's a perhaps slightly more illustrative mod of your example program, to show that the last image is not the last `to-draw` image.

```
include world
include image

big-bang(10,
  [list:
    last-image(lam(x): text(num-to-string(75), 75, "red") end),
    to-draw(lam(x): text(num-to-string(x), x, "blue") end),
    on-tick(lam(x): x + 10 end),
    stop-when(lam(x): x > 90 end)])
```

**Comment by schanzer** [comment]
Subsumed by #218

**Comment by schanzer** [comment]
Since this is obsolete (see #218), can we close?

**Comment by schanzer** [comment]
Paging @ds26gte

**Pull request closed without merging**